crates/sui-types/src/temporary_store.rs

```diff
@@ -689,6 +688,154 @@ impl<S> TemporaryStore<S> {
+                        }
+                        Owner::Immutable => unreachable!("Immutable objects cannot be deleted"),
+                    }
+                }
+                DeleteKind::UnwrapThenDelete => {
+                    // unwrapped-then-deleted object was not an input to the tx,
+                    // no ownership checks needed
+                    ()
+                }
+            }
+        }
+        Ok((objs_to_authenticate, authenticated_objs))
+    }
+
+    // check that every object read is owned directly or indirectly by sender, sponsor, or a shared object input
+    pub fn check_ownership_invariants(
+        &self,
+        sender: &SuiAddress,
+        gas: &[ObjectRef],
+    ) -> SuiResult<()> {
+        let (mut objects_to_authenticate, mut authenticated_objects) =
+            self.get_objects_to_authenticate(sender, gas)?;
+        // iterate until we we have authenticated all objects, or reached a fixed point with remaining objects to authenticate
+        loop {
+            assert_eq!(
+                objects_to_authenticate
+                    .intersection(&authenticated_objects)
+                    .count(),
+                0
+            );
+
+            let mut to_add = HashSet::new();
+            let mut to_remove: HashSet<ObjectID> = HashSet::new();
+            let mut progress = false;
+            for id in &objects_to_authenticate {
+                let old_obj = self.store.get_object(&id)?.unwrap();
+                if let Owner::ObjectOwner(parent) = &old_obj.owner {
+                    to_remove.insert(*id);
+                    let parent_id = ObjectID::from(*parent);
+                    if !authenticated_objects.contains(&parent_id) {
+                        to_add.insert(parent_id);
+                    } else {
+                        // already authenticated the parent, nothing further to check
+                        authenticated_objects.insert(*id);
+                    }
+                    progress = true
+                } else {
+                    unreachable!("Other than the roots, a valid ownership chain should always consist of object owners. {:?}'s owner is {:?}", id, old_obj.owner)
+                }
+            }
+            for id in to_add {
+                if !authenticated_objects.contains(&id) {
+                    objects_to_authenticate.insert(id);
+                }
+            }
+            for id in to_remove {
+                objects_to_authenticate.remove(&id);
+            }
+            if objects_to_authenticate.is_empty() {
+                return Ok(());
+            }
+            // continue iterating until we fail to add new objects to authenticate or grow the set of authenticated objects
+            if !progress {
+                panic!("Failed to authenticate {:?} objects: {:?}. Successfully authenticated objects are {:?}", objects_to_authenticate.len(), objects_to_authenticate, authenticated_objects)
+            }
```

Comment:
> what if something ends up in both sets via different paths, etc

I am nervous about this too, and do have an assertion that guards against it. But your algorithm does indeed look simpler--let me give it a shot.

Reply:
I tried this (and it does indeed look much better!), but got stuck because `BTreeMap::pop_first` is unstable until Rust 1.66. I can't think of an easy way to work around this, so am inclined to keep the algorithm as-is until we update Rust (though definitely open to suggestions).

Reply:
I think it should be fine to convert it into a vector (i.e. treat it like a stack)? We don't actually need to check whether something's already in there.